File: fix_credit_card_amounts.py
#!/usr/bin/env python3

import json
import os

def fix_credit_card_amounts(user_data):
    """Fix credit card transaction amounts to follow proper conventions"""
    
    # Fix primary credit card transactions
    if 'personalInfo' in user_data and 'creditCards' in user_data['personalInfo']:
        credit_cards = user_data['personalInfo']['creditCards']
        
        for card_type in ['primary', 'secondary']:
            if card_type in credit_cards and 'transactions' in credit_cards[card_type]:
                transactions = credit_cards[card_type]['transactions']
                
                for transaction in transactions:
                    # Only payments should be positive, all other transactions should be negative
                    if transaction.get('type') != 'payment' and transaction.get('amount', 0) > 0:
                        transaction['amount'] = -abs(transaction['amount'])
                    elif transaction.get('type') == 'payment' and transaction.get('amount', 0) < 0:
                        transaction['amount'] = abs(transaction['amount'])
                
                # Recalculate balance (negative means debt owed)
                total_spent = sum(abs(t['amount']) for t in transactions if t.get('type') != 'payment')
                total_paid = sum(t['amount'] for t in transactions if t.get('type') == 'payment')
                credit_cards[card_type]['balance'] = -(total_spent - total_paid)
    
    return user_data

# Process all sample data files
sample_data_dir = '/Users/rogier/bluemountainbank/sample-data'
files = [
    'alice_martinez.json',
    'bob_johnson.json', 
    'carol_smith.json',
    'david_wilson.json',
    'emma_brown.json',
    'frank_miller.json'
]

total_fixed = 0

for filename in files:
    file_path = os.path.join(sample_data_dir, filename)
    
    try:
        # Read the current data
        with open(file_path, 'r') as f:
            user_data = json.load(f)
        
        print(f"\nProcessing {filename}...")
        
        # Count transactions before fixing
        transactions_before = 0
        positive_transactions_before = 0
        
        if ('personalInfo' in user_data and 
            'creditCards' in user_data['personalInfo']):
            
            credit_cards = user_data['personalInfo']['creditCards']
            
            for card_type in ['primary', 'secondary']:
                if card_type in credit_cards and 'transactions' in credit_cards[card_type]:
                    transactions = credit_cards[card_type]['transactions']
                    transactions_before += len(transactions)
                    positive_transactions_before += len([
                        t for t in transactions 
                        if t.get('type') != 'payment' and t.get('amount', 0) > 0
                    ])
        
        # Fix the amounts
        fixed_data = fix_credit_card_amounts(user_data)
        
        # Count after fixing
        positive_transactions_after = 0
        
        if ('personalInfo' in fixed_data and 
            'creditCards' in fixed_data['personalInfo']):
            
            credit_cards = fixed_data['personalInfo']['creditCards']
            
            for card_type in ['primary', 'secondary']:
                if card_type in credit_cards and 'transactions' in credit_cards[card_type]:
                    transactions = credit_cards[card_type]['transactions']
                    positive_transactions_after += len([
                        t for t in transactions 
                        if t.get('type') != 'payment' and t.get('amount', 0) > 0
                    ])
        
        print(f"  Total transactions: {transactions_before}")
        print(f"  Positive non-payment transactions before: {positive_transactions_before}")
        print(f"  Positive non-payment transactions after: {positive_transactions_after}")
        print(f"  Fixed: {positive_transactions_before - positive_transactions_after} transactions")
        
        # Write the fixed data back
        with open(file_path, 'w') as f:
            json.dump(fixed_data, f, indent=2)
        
        total_fixed += (positive_transactions_before - positive_transactions_after)
        
        print(f"✅ {filename} updated successfully")
        
    except Exception as error:
        print(f"❌ Error processing {filename}: {error}")

print(f"\n🎉 Fix complete! Total transactions fixed: {total_fixed}")
print('\nCredit card transaction amounts now follow proper conventions:')
print('  • Purchases/charges: negative amounts (money owed)')
print('  • Payments: positive amounts (money paid)')

File: fix-credit-card-amounts.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Function to fix credit card transaction amounts
function fixCreditCardAmounts(userData) {
    // Fix primary credit card transactions
    if (userData.personalInfo?.creditCards?.primary?.transactions) {
        userData.personalInfo.creditCards.primary.transactions.forEach(transaction => {
            // Only payments should be positive, all other transactions should be negative
            if (transaction.type !== 'payment' && transaction.amount > 0) {
                transaction.amount = -Math.abs(transaction.amount);
            } else if (transaction.type === 'payment' && transaction.amount < 0) {
                transaction.amount = Math.abs(transaction.amount);
            }
        });
        
        // Recalculate balance (negative means debt owed)
        const totalSpent = userData.personalInfo.creditCards.primary.transactions
            .filter(t => t.type !== 'payment')
            .reduce((sum, t) => sum + Math.abs(t.amount), 0);
        
        const totalPaid = userData.personalInfo.creditCards.primary.transactions
            .filter(t => t.type === 'payment')
            .reduce((sum, t) => sum + t.amount, 0);
        
        userData.personalInfo.creditCards.primary.balance = -(totalSpent - totalPaid);
    }
    
    // Fix secondary credit card transactions if exists
    if (userData.personalInfo?.creditCards?.secondary?.transactions) {
        userData.personalInfo.creditCards.secondary.transactions.forEach(transaction => {
            if (transaction.type !== 'payment' && transaction.amount > 0) {
                transaction.amount = -Math.abs(transaction.amount);
            } else if (transaction.type === 'payment' && transaction.amount < 0) {
                transaction.amount = Math.abs(transaction.amount);
            }
        });
        
        // Recalculate balance
        const totalSpent = userData.personalInfo.creditCards.secondary.transactions
            .filter(t => t.type !== 'payment')
            .reduce((sum, t) => sum + Math.abs(t.amount), 0);
        
        const totalPaid = userData.personalInfo.creditCards.secondary.transactions
            .filter(t => t.type === 'payment')
            .reduce((sum, t) => sum + t.amount, 0);
        
        userData.personalInfo.creditCards.secondary.balance = -(totalSpent - totalPaid);
    }
    
    return userData;
}

// Get all sample data files
const sampleDataDir = '/Users/rogier/bluemountainbank/sample-data';
const files = [
    'alice_martinez.json',
    'bob_johnson.json',
    'carol_smith.json',
    'david_wilson.json',
    'emma_brown.json',
    'frank_miller.json'
];

let totalFixed = 0;

files.forEach(filename => {
    const filePath = path.join(sampleDataDir, filename);
    
    try {
        // Read the current data
        const rawData = fs.readFileSync(filePath, 'utf8');
        const userData = JSON.parse(rawData);
        
        console.log(`\nProcessing ${filename}...`);
        
        // Count transactions before fixing
        let transactionsBefore = 0;
        let positiveTransactionsBefore = 0;
        
        if (userData.personalInfo?.creditCards?.primary?.transactions) {
            transactionsBefore += userData.personalInfo.creditCards.primary.transactions.length;
            positiveTransactionsBefore += userData.personalInfo.creditCards.primary.transactions
                .filter(t => t.type !== 'payment' && t.amount > 0).length;
        }
        
        if (userData.personalInfo?.creditCards?.secondary?.transactions) {
            transactionsBefore += userData.personalInfo.creditCards.secondary.transactions.length;
            positiveTransactionsBefore += userData.personalInfo.creditCards.secondary.transactions
                .filter(t => t.type !== 'payment' && t.amount > 0).length;
        }
        
        // Fix the amounts
        const fixedData = fixCreditCardAmounts(userData);
        
        // Count after fixing
        let positiveTransactionsAfter = 0;
        
        if (fixedData.personalInfo?.creditCards?.primary?.transactions) {
            positiveTransactionsAfter += fixedData.personalInfo.creditCards.primary.transactions
                .filter(t => t.type !== 'payment' && t.amount > 0).length;
        }
        
        if (fixedData.personalInfo?.creditCards?.secondary?.transactions) {
            positiveTransactionsAfter += fixedData.personalInfo.creditCards.secondary.transactions
                .filter(t => t.type !== 'payment' && t.amount > 0).length;
        }
        
        console.log(`  Total transactions: ${transactionsBefore}`);
        console.log(`  Positive non-payment transactions before: ${positiveTransactionsBefore}`);
        console.log(`  Positive non-payment transactions after: ${positiveTransactionsAfter}`);
        console.log(`  Fixed: ${positiveTransactionsBefore - positiveTransactionsAfter} transactions`);
        
        // Write the fixed data back
        fs.writeFileSync(filePath, JSON.stringify(fixedData, null, 2));
        
        totalFixed += (positiveTransactionsBefore - positiveTransactionsAfter);
        
        console.log(`✅ ${filename} updated successfully`);
        
    } catch (error) {
        console.error(`❌ Error processing ${filename}:`, error.message);
    }
});

console.log(`\n🎉 Fix complete! Total transactions fixed: ${totalFixed}`);
console.log('\nCredit card transaction amounts now follow proper conventions:');
console.log('  • Purchases/charges: negative amounts (money owed)');
console.log('  • Payments: positive amounts (money paid)');
